File: sdk/typescript/src/light_client/verify.ts
/**
 * Light client verification helpers.
 *
 * Goals:
 *  - Hash a canonical header (CBOR) and perform basic invariants checks
 *  - Verify parent→child linkage (number increments, parentHash match)
 *  - Verify DA light-proofs against the header's daRoot
 *  - Provide a generic binary-Merkle proof verifier and allow custom verifiers
 *
 * Notes:
 *  - This module intentionally keeps consensus-specific rules minimal. It does
 *    not recompute PoW/PoIES acceptance or difficulty. It only checks static
 *    header fields, linkage, and (optionally) DA light proofs.
 *  - Header hashing uses deterministic CBOR encoding via utils/cbor.
 *  - DA proofs vary by deployment. We support a generic binary Merkle branch
 *    (SHA3-256) and allow injecting a custom verifier for NMT or other trees.
 */

import { encodeCanonicalCBOR } from '../utils/cbor'
import { sha3_256 } from '../utils/hash'
import { bytesToHex, hexToBytes, isHex } from '../utils/bytes'

// ──────────────────────────────────────────────────────────────────────────────
// Types
// ──────────────────────────────────────────────────────────────────────────────

/** Minimal header view required for hashing & linkage checks. */
export interface LightHeader {
  number: number               // block height
  parentHash: string           // 0x…
  stateRoot: string            // 0x…
  txRoot: string               // 0x…
  receiptsRoot: string         // 0x…
  daRoot?: string              // 0x… (optional if chain has no DA)
  chainId: number
  timestamp?: number
  // policy roots (optional, checked if provided in VerifyHeaderOptions)
  poiesPolicyRoot?: string
  algPolicyRoot?: string
  // extra fields are allowed and preserved in hashing (CBOR canonical map)
  [k: string]: unknown
}

/** Basic verification result with optional warnings/errors. */
export interface VerifyResult {
  ok: boolean
  hash: string
  errors: string[]
  warnings: string[]
}

/** Options for header verification. */
export interface VerifyHeaderOptions {
  expectedChainId?: number
  expectedPoiesPolicyRoot?: string
  expectedAlgPolicyRoot?: string
  // When true, enforce that critical roots are present and 0x-hex
  strictRoots?: boolean
}

/** Parent→child linkage result */
export interface LinkResult {
  ok: boolean
  errors: string[]
}

/** Binary Merkle branch step */
export interface MerkleStep {
  /** Sibling hash (0x…) */
  sibling: string
  /** Position of the *current* hash relative to sibling at this step */
  position: 'left' | 'right'
}

/** Binary Merkle proof for a single leaf */
export interface BinaryProof {
  /** Pre-hashed leaf (0x…) OR raw bytes to be hashed. Provide exactly one. */
  leafHash?: string
  leafData?: string | Uint8Array
  /** Steps from leaf → root */
  path: MerkleStep[]
}

/** DA light proof envelope (generic) */
export interface DALightProof {
  /** Commitment or identifier of the blob/data referenced (0x…) */
  commitment?: string
  /** Expected root to match the header's daRoot (0x…) */
  root: string
  /** One or more binary proofs proving sampled leaves into the root */
  binary?: BinaryProof[]
  /** Optional custom verifier hook (e.g., NMT range proofs) */
  verifyWith?: (p: DALightProof, header: LightHeader) => boolean | Promise<boolean>
  /** Additional, chain-specific fields */
  [k: string]: unknown
}

// ──────────────────────────────────────────────────────────────────────────────
// Header hashing & checks
// ──────────────────────────────────────────────────────────────────────────────

/**
 * Compute the canonical header hash = sha3_256(CBOR(header_map)).
 * The map must contain the fields as provided; CBOR encoding is deterministic.
 */
export function hashHeader(header: LightHeader): string {
  // Defensive clone so we don't mutate caller's object when normalizing hex
  const h: Record<string, any> = { ...header }

  // Normalize critical roots to lowercase 0x-hex (if present)
  for (const k of ['parentHash', 'stateRoot', 'txRoot', 'receiptsRoot', 'daRoot', 'poiesPolicyRoot', 'algPolicyRoot']) {
    if (h[k] != null) h[k] = normalizeHex(h[k])
  }

  const cbor = encodeCanonicalCBOR(h)
  return '0x' + bytesToHex(sha3_256(cbor))
}

/**
 * Verify a single header's basic invariants.
 * Does not check PoW/PoIES acceptance—only static sanity and configured expectations.
 */
export function verifyHeader(header: LightHeader, opts: VerifyHeaderOptions = {}): VerifyResult {
  const errors: string[] = []
  const warnings: string[] = []

  // Field sanity
  if (!Number.isFinite(header.number) || header.number < 0) errors.push('invalid header.number')
  for (const k of ['parentHash', 'stateRoot', 'txRoot', 'receiptsRoot'] as const) {
    if (!isHex(header[k] as string)) errors.push(`invalid ${k} (expected 0x-hex)`)
  }
  if (opts.strictRoots && header.daRoot == null) errors.push('daRoot missing (strictRoots)')
  if (header.daRoot != null && !isHex(header.daRoot)) errors.push('invalid daRoot (expected 0x-hex)')

  if (opts.expectedChainId != null && header.chainId !== opts.expectedChainId) {
    errors.push(`chainId mismatch: header=${header.chainId} expected=${opts.expectedChainId}`)
  }
  if (opts.expectedPoiesPolicyRoot) {
    const h = header.poiesPolicyRoot ? normalizeHex(header.poiesPolicyRoot) : ''
    const exp = normalizeHex(opts.expectedPoiesPolicyRoot)
    if (h !== exp) errors.push('poiesPolicyRoot mismatch')
  }
  if (opts.expectedAlgPolicyRoot) {
    const h = header.algPolicyRoot ? normalizeHex(header.algPolicyRoot) : ''
    const exp = normalizeHex(opts.expectedAlgPolicyRoot)
    if (h !== exp) errors.push('algPolicyRoot mismatch')
  }

  const hash = hashHeader(header)
  return { ok: errors.length === 0, hash, errors, warnings }
}

/**
 * Verify parent→child linkage: child.number = parent.number+1 and
 * child.parentHash == hashHeader(parent).
 */
export function verifyLink(parent: LightHeader, child: LightHeader): LinkResult {
  const errors: string[] = []
  const parentHash = hashHeader(parent)
  if (child.number !== parent.number + 1) {
    errors.push(`height does not increment: parent=${parent.number} child=${child.number}`)
  }
  if (normalizeHex(child.parentHash) !== parentHash) {
    errors.push('parentHash mismatch between child and hashed parent')
  }
  return { ok: errors.length === 0, errors }
}

// ──────────────────────────────────────────────────────────────────────────────
/* DA light-proof verification (binary Merkle by default) */
// ──────────────────────────────────────────────────────────────────────────────

export interface VerifyDAResult {
  ok: boolean
  errors: string[]
  checked: number    // number of leaf proofs checked
}

/**
 * Verify a DA light proof against the header's daRoot.
 * - Uses built-in binary-Merkle verification for `proof.binary` items
 * - If `proof.verifyWith` is provided, it is invoked (async allowed) and must return true
 */
export async function verifyDA(proof: DALightProof, header: LightHeader): Promise<VerifyDAResult> {
  const errors: string[] = []
  const expectedRoot = normalizeHex(header.daRoot ?? '')
  if (!expectedRoot) return { ok: false, errors: ['header lacks daRoot'], checked: 0 }

  if (!isHex(proof.root)) errors.push('proof.root is not hex')
  const root = normalizeHex(proof.root)
  if (root !== expectedRoot) errors.push('proof root does not match header.daRoot')

  // Built-in binary proofs (optional)
  let checked = 0
  if (Array.isArray(proof.binary) && proof.binary.length) {
    for (const p of proof.binary) {
      try {
        const ok = verifyBinaryMerkle(p, root)
        if (!ok) errors.push('binary Merkle proof failed')
        else checked++
      } catch (e: any) {
        errors.push(`binary Merkle verification error: ${e?.message || String(e)}`)
      }
    }
  }

  // Custom verifier hook (e.g., NMT range proofs). If provided, it must pass.
  if (proof.verifyWith) {
    try {
      const ok = await proof.verifyWith(proof, header)
      if (!ok) errors.push('custom DA verifier reported failure')
    } catch (e: any) {
      errors.push(`custom DA verifier threw: ${e?.message || String(e)}`)
    }
  }

  return { ok: errors.length === 0, errors, checked }
}

/**
 * Verify a binary Merkle proof path: starting from leaf, combine with siblings up to root.
 * Hash function: sha3_256; combine(left,right) = sha3_256(0x01 || left || right)
 * The 0x01 domain tag reduces cross-protocol collisions.
 */
export function verifyBinaryMerkle(p: BinaryProof, expectedRoot: string): boolean {
  const root = normalizeHex(expectedRoot)
  if (!p || !Array.isArray(p.path) || p.path.length === 0) throw new Error('empty path')
  let h = p.leafHash ? normalizeHex(p.leafHash) : undefined
  if (!h) {
    if (p.leafData == null) throw new Error('leafData or leafHash required')
    const data = typeof p.leafData === 'string' ? hexOrUtf8ToBytes(p.leafData) : p.leafData
    h = '0x' + bytesToHex(sha3_256(concat(domain(0x00), data))) // 0x00 domain for leaf
  }
  let cur = hexToBytes(h)
  for (const step of p.path) {
    const sib = hexToBytes(normalizeHex(step.sibling))
    if (step.position === 'left') {
      cur = sha3_256(concat(domain(0x01), cur, sib))
    } else if (step.position === 'right') {
      cur = sha3_256(concat(domain(0x01), sib, cur))
    } else {
      throw new Error(`invalid step.position: ${String(step.position)}`)
    }
  }
  const got = '0x' + bytesToHex(cur)
  return got.toLowerCase() === root
}

// ──────────────────────────────────────────────────────────────────────────────
// Bundled flow convenience
// ──────────────────────────────────────────────────────────────────────────────

export interface VerifyBundleOptions extends VerifyHeaderOptions {
  daProof?: DALightProof
  parent?: LightHeader
}

/** Verify header (+ optional parent link) and optional DA proof in one call. */
export async function verifyBundle(header: LightHeader, opts: VerifyBundleOptions = {}) {
  const hdr = verifyHeader(header, opts)
  const link = opts.parent ? verifyLink(opts.parent, header) : { ok: true, errors: [] as string[] }
  const da = opts.daProof ? await verifyDA(opts.daProof, header) : { ok: true, errors: [] as string[], checked: 0 }
  const ok = hdr.ok && link.ok && da.ok
  return { ok, header: hdr, link, da }
}

// ──────────────────────────────────────────────────────────────────────────────
/* Utils */
// ──────────────────────────────────────────────────────────────────────────────

function normalizeHex(h: string): string {
  if (typeof h !== 'string') throw new Error('expected hex string')
  const s = h.startsWith('0x') || h.startsWith('0X') ? h : ('0x' + h)
  if (!isHex(s)) throw new Error(`invalid hex: ${h}`)
  return s.toLowerCase()
}

function hexOrUtf8ToBytes(x: string): Uint8Array {
  return isHex(x) ? hexToBytes(x) : new TextEncoder().encode(x)
}

function bytesToHexLower(b: Uint8Array): string {
  return bytesToHex(b).toLowerCase()
}

function domain(tag: number): Uint8Array {
  // Single-byte domain tags: 0x00 = leaf, 0x01 = node
  const u = new Uint8Array(1); u[0] = tag & 0xff; return u
}

function concat(...parts: Uint8Array[]): Uint8Array {
  const n = parts.reduce((a, p) => a + p.length, 0)
  const out = new Uint8Array(n)
  let off = 0
  for (const p of parts) { out.set(p, off); off += p.length }
  return out
}

// Re-exports for convenience
export default {
  hashHeader,
  verifyHeader,
  verifyLink,
  verifyDA,
  verifyBinaryMerkle,
  verifyBundle
}

File: sdk/rust/src/light_client/verify.rs
//! Light client verification: header linkage and DA light proofs.
//!
//! This module performs **offline** checks suitable for wallet/SDK usage:
//! - Header chain linkage: `parentHash` continuity, monotonic height.
//! - Optional DA light proof verification against a header's `daRoot`.
//!
//! It intentionally avoids full consensus validation. For production,
//! pair this with trusted checkpoints (anchor headers) and server-side
//! SPV/consensus checks where applicable.

use hex::FromHex;
use serde::{Deserialize, Serialize};
use serde_json::Value as JsonValue;
use sha3::{Digest, Sha3_256};
use std::collections::BTreeMap;

/// Minimal header view (as returned by RPC explorers or node light endpoints).
/// Unknown fields are preserved in `extra` for forward compatibility.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct LightHeader {
    pub hash: String,         // 0x-prefixed
    pub parent_hash: String,  // 0x-prefixed
    #[serde(alias = "height", alias = "number")]
    pub number: u64,
    #[serde(default)]
    pub da_root: Option<String>, // 0x-prefixed (NMT root)
    #[serde(default)]
    pub state_root: Option<String>,
    #[serde(default)]
    pub receipts_root: Option<String>,
    #[serde(default)]
    pub chain_id: Option<u64>,
    #[serde(flatten)]
    pub extra: BTreeMap<String, JsonValue>,
}

/// A single Merkle sample proof for availability.
/// Branch is bottom-up; each node is a 32-byte 0x-hex digest.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SampleProof {
    pub index: u64,
    pub leaf_hash: String,       // 0x-hex, 32 bytes
    pub branch: Vec<String>,     // 0x-hex digests, deepest-first
}

/// Availability/light proof bundle (subset that light clients need).
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AvailabilityProof {
    pub samples: Vec<SampleProof>,
    #[serde(default)]
    pub leaves_count: Option<u64>,
    #[serde(default)]
    pub algo: Option<String>, // e.g., "sha3-256"
    #[serde(flatten)]
    pub extra: BTreeMap<String, JsonValue>,
}

/// Summary of performed checks.
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct VerifySummary {
    pub headers_checked: usize,
    pub da_samples_checked: usize,
}

/// Public API: verify a contiguous header slice (oldest → newest).
/// If `anchor_hash` is provided, require headers[0].parent_hash == anchor_hash.
/// Returns a summary of performed checks.
pub fn verify_headers(headers: &[LightHeader], anchor_hash: Option<&str>) -> Result<VerifySummary, String> {
    if headers.is_empty() {
        return Err("verify_headers: empty slice".into());
    }
    // Verify anchor linkage if provided.
    if let Some(anchor) = anchor_hash {
        if !eq_hex_case_insensitive(&headers[0].parent_hash, anchor) {
            return Err(format!(
                "anchor mismatch: {} (header[0].parentHash) != {} (anchor)",
                headers[0].parent_hash, anchor
            ));
        }
    }
    // Verify continuity and monotonic height.
    for i in 1..headers.len() {
        let prev = &headers[i - 1];
        let cur = &headers[i];
        if !eq_hex_case_insensitive(&cur.parent_hash, &prev.hash) {
            return Err(format!(
                "parent linkage break at index {}: parentHash={} != prev.hash={}",
                i, cur.parent_hash, prev.hash
            ));
        }
        if cur.number != prev.number + 1 {
            return Err(format!(
                "height not monotonic at index {}: {} then {}",
                i, prev.number, cur.number
            ));
        }
    }
    Ok(VerifySummary {
        headers_checked: headers.len(),
        da_samples_checked: 0,
    })
}

/// Verify a DA availability proof against a 32-byte `da_root` (0x-hex).
/// Returns the number of samples checked on success.
pub fn verify_da_proof(da_root: &str, proof: &AvailabilityProof) -> Result<usize, String> {
    let root = hex32(da_root)?;
    if let Some(algo) = &proof.algo {
        let a = algo.to_ascii_lowercase();
        if a != "sha3-256" && a != "sha3_256" {
            return Err(format!("unsupported algo: {}", algo));
        }
    }
    if proof.samples.is_empty() {
        return Err("availability proof has zero samples".into());
    }
    for (idx, sp) in proof.samples.iter().enumerate() {
        // Accept either full leaf hash or a leaf preimage already hashed by producer.
        let mut cur = hex32(&sp.leaf_hash)?;
        let mut index = sp.index as usize;

        for sib_hex in &sp.branch {
            let sib = hex32(sib_hex)?;
            let (left, right) = if (index & 1) == 0 {
                (cur, sib)
            } else {
                (sib, cur)
            };
            cur = merkle_parent_sha3_256(&left, &right);
            index >>= 1;
        }
        if cur != root {
            return Err(format!(
                "sample {} failed: computed root {} != expected {}",
                idx,
                hex::encode_prefixed(cur),
                hex::encode_prefixed(root)
            ));
        }
    }
    Ok(proof.samples.len())
}

/// Convenience: verify headers + (optional) DA proof against the **last** header's daRoot.
pub fn verify_headers_and_da(
    headers: &[LightHeader],
    anchor_hash: Option<&str>,
    da_proof: Option<&AvailabilityProof>,
) -> Result<VerifySummary, String> {
    let mut sum = verify_headers(headers, anchor_hash)?;
    if let Some(p) = da_proof {
        let last = headers.last().ok_or("internal: empty headers")?;
        let da_root = last
            .da_root
            .as_ref()
            .ok_or("last header has no daRoot to verify against")?;
        let n = verify_da_proof(da_root, p)?;
        sum.da_samples_checked = n;
    }
    Ok(sum)
}

// --------------------------- Internal helpers -------------------------------

fn eq_hex_case_insensitive(a: &str, b: &str) -> bool {
    trim_0x(a).eq_ignore_ascii_case(trim_0x(b))
}

fn trim_0x(s: &str) -> &str {
    s.strip_prefix("0x").unwrap_or(s)
}

fn hex32(s: &str) -> Result<[u8; 32], String> {
    let s = trim_0x(s);
    let bytes = <Vec<u8>>::from_hex(s).map_err(|e| format!("hex decode: {e}"))?;
    if bytes.len() != 32 {
        return Err(format!("expected 32 bytes, got {}", bytes.len()));
    }
    let mut out = [0u8; 32];
    out.copy_from_slice(&bytes);
    Ok(out)
}

fn merkle_parent_sha3_256(left: &[u8; 32], right: &[u8; 32]) -> [u8; 32] {
    let mut hasher = Sha3_256::new();
    hasher.update(left);
    hasher.update(right);
    let out = hasher.finalize();
    let mut arr = [0u8; 32];
    arr.copy_from_slice(&out);
    arr
}

// ------------------------------- Tests --------------------------------------

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn header_linkage_ok() {
        let h1 = LightHeader {
            hash: "0x11".into(),
            parent_hash: "0x00".into(),
            number: 1,
            da_root: None,
            state_root: None,
            receipts_root: None,
            chain_id: Some(1),
            extra: BTreeMap::new(),
        };
        let h2 = LightHeader {
            hash: "0x22".into(),
            parent_hash: "0x11".into(),
            number: 2,
            da_root: None,
            state_root: None,
            receipts_root: None,
            chain_id: Some(1),
            extra: BTreeMap::new(),
        };
        let s = verify_headers(&[h1, h2], Some("0x00")).unwrap();
        assert_eq!(s.headers_checked, 2);
        assert_eq!(s.da_samples_checked, 0);
    }

    #[test]
    fn header_linkage_breaks() {
        let h1 = LightHeader {
            hash: "0x11".into(),
            parent_hash: "0x00".into(),
            number: 1,
            da_root: None,
            state_root: None,
            receipts_root: None,
            chain_id: None,
            extra: BTreeMap::new(),
        };
        let h2 = LightHeader {
            hash: "0x22".into(),
            parent_hash: "0x99".into(), // wrong
            number: 2,
            da_root: None,
            state_root: None,
            receipts_root: None,
            chain_id: None,
            extra: BTreeMap::new(),
        };
        let err = verify_headers(&[h1, h2], None).unwrap_err();
        assert!(err.contains("parent linkage break"));
    }

    #[test]
    fn merkle_sample_verifies() {
        // Build a tiny Merkle tree with two leaves: L0, L1 (already 32-byte digests for simplicity).
        let l0 = Sha3_256::digest(b"leaf0");
        let l1 = Sha3_256::digest(b"leaf1");
        let mut l0_arr = [0u8; 32];
        let mut l1_arr = [0u8; 32];
        l0_arr.copy_from_slice(&l0);
        l1_arr.copy_from_slice(&l1);
        let root = merkle_parent_sha3_256(&l0_arr, &l1_arr);
        let root_hex = hex::encode_prefixed(root);

        // Proof for index 0: branch contains sibling (l1), index=0 so current is left.
        let proof = AvailabilityProof {
            samples: vec![SampleProof {
                index: 0,
                leaf_hash: hex::encode_prefixed(l0_arr),
                branch: vec![hex::encode_prefixed(l1_arr)],
            }],
            leaves_count: Some(2),
            algo: Some("sha3-256".into()),
            extra: BTreeMap::new(),
        };

        let checked = verify_da_proof(&root_hex, &proof).unwrap();
        assert_eq!(checked, 1);
    }

    #[test]
    fn merkle_sample_fails_on_wrong_root() {
        let bogus_root = "0x".to_string() + &"00".repeat(32);
        let sample = SampleProof {
            index: 0,
            leaf_hash: "0x".to_string() + &"00".repeat(32),
            branch: vec!["0x".to_string() + &"00".repeat(32)],
        };
        let proof = AvailabilityProof {
            samples: vec![sample],
            leaves_count: None,
            algo: None,
            extra: BTreeMap::new(),
        };
        assert!(verify_da_proof(&bogus_root, &proof).is_err());
    }

    #[test]
    fn json_roundtrip_structs() {
        let j = json!({
            "hash":"0x01",
            "parentHash":"0x00",
            "number": 7,
            "daRoot": "0x" + &"11".repeat(32),
            "unknownField": 42
        });
        let h: LightHeader = serde_json::from_value(j).unwrap();
        assert_eq!(h.number, 7);
        assert!(h.extra.contains_key("unknownField"));
    }
}
